### PR TITLE
Get process list, submit vote

### PR DESCRIPTION
**Pull request by NateWilliams2**

### CODE REVIEW

**Comment by NateWilliams2** [comment]
To be merged on top of #10

File: types/api.go [hunk 1]
```diff
@@ -75,6 +79,20 @@ type APIProcess struct {
...
 }
 
+type APIElection struct {
+	OrgEthAddress   types.HexBytes `json:"orgEthAddress,omitempty" db:"organization_eth_address"`
+	ElectionID      types.HexBytes `json:"electionId,omitempty" db:"process_id"`
+	Title           string         `json:"title,omitempty" db:"title"`
+	CensusID        string         `json:"censusId,omitempty" db:"census_id"`
+	StartDate       time.Time      `json:"startDate,omitempty" db:"start_date"`
+	EndDate         time.Time      `json:"endDate,omitempty" db:"end_date"`
+	StartBlock      int            `json:"startBlock,omitempty" db:"start_block"`
```
Comment (p4u) [inline types/api.go:89]: I think uint32 or uint64 are better for block numbers (since a negative block does not makes sense)

File: urlapi/entity.go [hunk 2]
```diff
@@ -516,6 +525,88 @@ func (u *URLAPI) createProcessHandler(msg *bearerstdapi.BearerStandardAPIdata,
...
+	var integratorPrivKey []byte
+	var err error
+	var resp []types.APIElection
+
+	if integratorPrivKey, entityId, _, err = u.authEntityPermissions(msg, ctx); err != nil {
+		log.Error(err)
+		return err
+	}
+
+	path := strings.Split(ctx.Request.URL.Path, "/")
```
Comment (p4u) [inline urlapi/entity.go:546]: Why not declaring this as a URL parameter such as: `https://server/v1/priv/organizations/<organizationId>/elections/<type>`

In general I think a handler should never access directly to `URL.Path`

File: urlapi/entity.go [hunk 3]
```diff
@@ -516,6 +525,88 @@ func (u *URLAPI) createProcessHandler(msg *bearerstdapi.BearerStandardAPIdata,
...
+	pathSuffix := path[len(path)-1]
+	switch pathSuffix {
+	case "active", "ended", "upcoming":
+		var tempProcessList []string
+		var totalProcs int
+		var currentHeight uint32
+		if currentHeight, err = u.vocClient.GetCurrentBlock(); err != nil {
+			log.Errorf("could not get current block height: %v", err)
```
Comment (p4u) [inline urlapi/entity.go:554]: Let's not log an error if it is already returned (layer below would choose what to do with it, if printing it or not)

File: urlapi/entity.go [hunk 4]
```diff
@@ -516,6 +525,88 @@ func (u *URLAPI) createProcessHandler(msg *bearerstdapi.BearerStandardAPIdata,
...
+			return fmt.Errorf("could not get current block height: %v", err)
```
Comment (p4u) [inline urlapi/entity.go:555]: Please, wrap the errors correctly with `%w`

Also, wrapping the errors is useful to understand where it was originated. Somethod like `listProcessesPrivateHandler: cannot get current block height: %w`

File: urlapi/entity.go [hunk 5]
```diff
@@ -516,6 +525,88 @@ func (u *URLAPI) createProcessHandler(msg *bearerstdapi.BearerStandardAPIdata,
...
+		}
+		cont := true
+		for cont {
+			if tempProcessList, err = u.vocClient.GetProcessList(entityId,
+				"", "", "", 0, false, totalProcs, 64); err != nil {
+				log.Errorf("%s not a valid request path", ctx.Request.URL.Path)
+				return fmt.Errorf("%s not a valid request path", ctx.Request.URL.Path)
+			}
+			if len(tempProcessList) < 64 {
```
Comment (p4u) [inline urlapi/entity.go:564]: No need to declare cont. 

If cont is false, do we want to keep runing the loop code or should it just break the loop inmediatly? If so, use `break`. 

If you really need keep executing code if cont=false, just use `len(tempProcessList) < 64` as loop condition

**Review by p4u — CHANGES_REQUESTED** [review]

File: urlapi/public.go [hunk 6]
```diff
@@ -84,10 +95,67 @@ func (u *URLAPI) listProcessesHandler(msg *bearerstdapi.BearerStandardAPIdata, c
...
-			return fmt.Errorf("%s not a valid request path", ctx.Request.URL.Path)
+	case "active", "ended", "upcoming":
+		var tempProcessList []string
+		var totalProcs int
+		var currentHeight uint32
+		if currentHeight, err = u.vocClient.GetCurrentBlock(); err != nil {
+			log.Errorf("could not get current block height: %v", err)
+			return fmt.Errorf("could not get current block height: %v", err)
+		}
+		cont := true
```
Comment (p4u) [inline urlapi/public.go:106]: same code here? I think this method requires also a refactor, its hard to read

File: urlapi/entity.go (hunk 4 above)
Comment (NateWilliams2) [reply urlapi/entity.go:555]: Do we want to wrap errors returned over HTTP? Imo an integrator getting a message starting with "listProcessesPrivateHandler" provides no meaningful information to them, even though it's helpful to us. I'll work on wrapping these error messages though, for the time being.

Comment (p4u) [reply urlapi/entity.go]: Umm, ok you'r right. Let's not add the method name but still use `%w`

File: urlapi/entity.go [hunk 7]
```diff
@@ -620,10 +609,30 @@ func (u *URLAPI) authEntityPermissions(msg *bearerstdapi.BearerStandardAPIdata,
 		return nil, nil, nil, err
 	}
 	if organization, err = u.db.GetOrganization(integratorPrivKey, entityID); err != nil {
-		return nil, nil, nil, fmt.Errorf("entity %X could not be fetched from the db: %v", entityID, err)
+		return nil, nil, nil, fmt.Errorf("entity %X could not be fetched from the db: %w", entityID, err)
```
Comment (p4u) [inline urlapi/entity.go:612]: I know this is not a change from this PR, but it would be much friendly if authEntityPermission returns a struct ` {integratorPrivKey,entityID, organization}` instead of 3 unamed parameters.

File: urlapi/entity.go [hunk 8]
```diff
@@ -620,10 +609,30 @@ func (u *URLAPI) authEntityPermissions(msg *bearerstdapi.BearerStandardAPIdata,
...
+		StartDate:       election.StartDate,
+		EndDate:         election.EndDate,
+		StartBlock:      uint32(election.StartBlock),
+		EndBlock:        uint32(election.EndBlock),
+		Confidential:    election.Confidential,
+		HiddenResults:   election.HiddenResults,
+		MetadataPrivKey: election.MetadataPrivKey,
+	}
+	if election.CensusID.UUID == uuid.Nil {
+		newElection.CensusID = ""
```
Comment (p4u) [inline urlapi/entity.go:635]: can you add a coment to clarify why this field rewrite?

File: urlapi/public.go [hunk 9]
```diff
@@ -235,21 +239,21 @@ func (u *URLAPI) parseProcessInfo(vc *indexertypes.Process,
 	if results != nil {
 		process.VoteCount = results.Height
 		if process.Results, err = aggregateResults(meta, results); err != nil {
-			log.Errorf("could not aggregate results: %v", err)
+			log.Error(fmt.Errorf("could not aggregate results: %w", err))
```
Comment (p4u) [inline urlapi/public.go:242]: If what you want is to log a message, then use `%v`instead of `%w`

File: urlapi/public.go [hunk 10]
```diff
@@ -191,10 +175,30 @@ func (u *URLAPI) getOrganizationHandler(msg *bearerstdapi.BearerStandardAPIdata,
...
+	}
+
+	return sendResponse(resp, ctx)
+}
+
 // TODO add listProcessesInfoHandler
 
 func (u *URLAPI) parseProcessInfo(vc *indexertypes.Process,
-	results *types.VochainResults, meta *types.ProcessMetadata) (process types.APIProcess) {
+	results *types.VochainResults, meta *types.ProcessMetadata) (process types.APIElectionInfo) {
```
Comment (p4u) [inline urlapi/public.go:201]: Please, in general do not use this syntax (define the var on the method signature). It makes code harded to read specially for big functions like this one.

Also try to aggregate as maximum values as possible when creating the var, such as:
```
process := types.APIElectionInfo{ 
 Type: "blind-",
 Title : meta.Title["default"],
 Description : meta.Description["default"],
 }
```

File: urlapi/public.go [hunk 11]
```diff
@@ -235,21 +239,21 @@ func (u *URLAPI) parseProcessInfo(vc *indexertypes.Process,
...
 	process.ElectionID = vc.ID
 
-	process.StartBlock = fmt.Sprintf("%d", vc.StartBlock)
-	process.EndBlock = fmt.Sprintf("%d", vc.EndBlock)
+	process.StartBlock = vc.StartBlock
+	process.EndBlock = vc.EndBlock
 
 	if process.StartDate, err = u.estimateBlockTime(vc.StartBlock); err != nil {
-		log.Warnf("could not estimate startDate at %d: %v", vc.StartBlock, err)
+		log.Warnf("could not estimate startDate at %d: %s", vc.StartBlock, err.Error())
```
Comment (p4u) [inline urlapi/public.go:252]: `%v` is fine for printing err

**Review by p4u — CHANGES_REQUESTED** [review]

File: urlapi/public.go [hunk 12]
```diff
@@ -357,6 +361,89 @@ func (u *URLAPI) estimateBlockHeight(target time.Time) (uint32, error) {
 	return currentHeight + uint32(blockDiff), nil
 }
 
+func (u *URLAPI) getProcessList(filter string, integratorPrivKey, entityId []byte,
+	private bool) (pub []types.APIElectionSummary, priv []types.APIElectionSummary, err error) {
```
Comment (p4u) [inline urlapi/public.go]: same as before regarding signature vars

File: urlapi/public.go [hunk 13]
```diff
@@ -357,6 +361,89 @@ func (u *URLAPI) estimateBlockHeight(target time.Time) (uint32, error) {
...
+		var fullProcessList []string
+		var currentHeight uint32
+		if currentHeight, err = u.vocClient.GetCurrentBlock(); err != nil {
+			return nil, nil, fmt.Errorf("could not get current block height: %w", err)
+		}
+		// loop to fetch all processes
+		for {
+			if tempProcessList, err = u.vocClient.GetProcessList(entityId,
+				"", "", "", 0, false, len(fullProcessList), 64); err != nil {
+				return nil, nil, fmt.Errorf("%s not a valid filter", filter)
```
Comment (p4u) [inline urlapi/public.go]: filter here is either `active, ended, upcoming` so it cannot be wrong, right? if `GetProcessList()` returns an error is not because of the filter

File: urlapi/public.go [hunk 14]
```diff
@@ -357,6 +361,89 @@ func (u *URLAPI) estimateBlockHeight(target time.Time) (uint32, error) {
...
+			if len(tempProcessList) < 64 {
+				break
+			}
+		}
+		// loop to fetch processes from db, filter by date
+		for _, processID := range fullProcessList {
+			var processIDBytes []byte
+			var newProcess *types.Election
+			if processIDBytes, err = hex.DecodeString(processID); err != nil {
+				log.Error(err)
```
Comment (p4u) [inline urlapi/public.go]: add some extra info to this message so it will be easier to find

File: urlapi/public.go [hunk 15]
```diff
@@ -357,6 +361,89 @@ func (u *URLAPI) estimateBlockHeight(target time.Time) (uint32, error) {
...
+				continue
+			}
+			if private {
+				newProcess, err = u.db.GetElection(integratorPrivKey,
+					entityId, processIDBytes)
+			} else {
+				newProcess, err = u.db.GetElectionPublic(entityId, processIDBytes)
+			}
+			if err != nil {
+				log.Warn(fmt.Errorf("could not get election,"+
```
Comment (p4u) [inline urlapi/public.go]: use `%v`and remove fmt.Errorf()

File: urlapi/public.go [hunk 16]
```diff
@@ -357,6 +361,89 @@ func (u *URLAPI) estimateBlockHeight(target time.Time) (uint32, error) {
...
+		return nil, nil, fmt.Errorf("filter %s unimplemented", filter)
+	default:
+		return nil, nil, fmt.Errorf("%s not a valid filter", filter)
+
+	}
+	return priv, pub, nil
+}
+
+func appendProcess(priv, pub []types.APIElectionSummary, newProcess *types.Election,
+	private bool) (privateElections []types.APIElectionSummary,
```
Comment (p4u) [inline urlapi/public.go]: same

File: urlapi/public.go [hunk 17]
```diff
@@ -357,6 +361,89 @@ func (u *URLAPI) estimateBlockHeight(target time.Time) (uint32, error) {
...
+	publicElections []types.APIElectionSummary) {
+	if private {
+		priv = append(priv, reflectElectionPrivate(*newProcess))
```
Comment (p4u) [inline urlapi/public.go]: `return append(priv, reflectElectionPrivate(*newProcess), nil` looks better to me (and remove the else case)

File: urlapi/public.go [hunk 18]
```diff
@@ -357,6 +361,89 @@ func (u *URLAPI) estimateBlockHeight(target time.Time) (uint32, error) {
...
+	} else {
+		if newProcess.Confidential {
+			priv = append(priv, reflectElectionPublic(*newProcess))
```
Comment (p4u) [inline urlapi/public.go]: `return append(priv, reflectElectionPublic(*newProcess), nil`

File: urlapi/public.go [hunk 19]
```diff
@@ -357,6 +361,89 @@ func (u *URLAPI) estimateBlockHeight(target time.Time) (uint32, error) {
...
+		} else {
+			pub = append(pub, reflectElectionPublic(*newProcess))
+		}
+	}
+	return priv, pub
```
Comment (p4u) [inline urlapi/public.go]: `return nil, append(pub, reflectElectionPublic(*newProcess))`

Please, add comments to the returns

File: vocclient/vocclient.go [hunk 20]
```diff
@@ -210,7 +216,7 @@ func (c *Client) FetchProcessMetadata(URI string) (process *apitypes.ProcessMeta
 	return process, nil
 }
 
-func (c *Client) FetchOrganizationMetadata(URI string) (entity *apitypes.EntityMetadata, _ error) {
+func (c *Client) FetchOrganizationMetadata(URI string) (entity *types.EntityMetadata, _ error) {
```
Comment (p4u) [inline vocclient/vocclient.go:219]: upss... why all these var definitions in function signature? I don't see the point sorry :(

File: vocclient/vocclient.go [hunk 21]
```diff
@@ -433,3 +439,63 @@ func (c *Client) CreateProcess(
 	}
 	return p.Process.StartBlock, nil
 }
+
+func (c *Client) RelayTx(reqBody []byte) (string, error) {
+	var err error
+
```
Comment (p4u) [inline vocclient/vocclient.go:445]: blank line

File: vocclient/vocclient.go [hunk 22]
```diff
@@ -433,3 +439,63 @@ func (c *Client) CreateProcess(
...
+	var gw Gateway
+	if gw, err = c.pool.activeGateway(); err != nil {
+		return "", err
+	}
+
+	var reqOuter jsonrpcapi.RequestMessage
+	if err := json.Unmarshal(reqBody, &reqOuter); err != nil {
+		return "", fmt.Errorf("%s: %v", "sumbitRawTx", err)
```
Comment (p4u) [inline vocclient/vocclient.go:453]: `fmt.Errorf("submitwarTx error: %w", err)`

File: vocclient/vocclient.go [hunk 23]
```diff
@@ -433,3 +439,63 @@ func (c *Client) CreateProcess(
...
+	}
+	reqId := reqOuter.ID
+
+	message := []byte{}
+	if gw.client.WS != nil {
+		tctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
```
Comment (p4u) [inline vocclient/vocclient.go:459]: make this 1*time.Minute a constant (better try to reduce magic numbers at minimum)

File: vocclient/vocclient.go [hunk 24]
```diff
@@ -433,3 +439,63 @@ func (c *Client) CreateProcess(
...
+	}
+
+	var reqOuter jsonrpcapi.RequestMessage
+	if err := json.Unmarshal(reqBody, &reqOuter); err != nil {
+		return "", fmt.Errorf("%s: %v", "sumbitRawTx", err)
+	}
+	reqId := reqOuter.ID
+
+	message := []byte{}
+	if gw.client.WS != nil {
```
Comment (p4u) [inline vocclient/vocclient.go:458]: Are we using Websockets? The httpRouter does not provide WS support AFAIK, so the Gateway connection won't work. Also we moved away in general from WS in favor of HTTP

File: vocclient/vocclient.go [hunk 25]
```diff
@@ -433,3 +439,63 @@ func (c *Client) CreateProcess(
...
+		if err != nil {
+			return "", err
+		}
+		message, err = io.ReadAll(resp.Body)
+		if err != nil {
+			return "", err
+		}
+		resp.Body.Close()
+	}
+	log.Debugf("response: %s", message)
```
Comment (p4u) [inline vocclient/vocclient.go:480]: response to what?

File: urlapi/util.go [hunk 26]
```diff
@@ -0,0 +1,380 @@
...
+		newQuestion := types.Question{
+			Title:       question.Title["default"],
+			Description: question.Description["default"],
+		}
+		for _, choice := range question.Choices {
+			newQuestion.Choices = append(newQuestion.Choices, choice.Title["default"])
+		}
+		process.Questions = append(process.Questions, newQuestion)
+	}
+	process.Status = strings.ToTitle(models.ProcessStatus_name[vc.Status])[0:1] +
```
Comment (p4u) [inline urlapi/util.go:141]: we should check `models.ProcessStatus_name[vc.Status]`before accessing to a specific position (it could be empty?)

File: urlapi/util.go [hunk 27]
```diff
@@ -0,0 +1,380 @@
...
+		strings.ToLower(models.ProcessStatus_name[vc.Status])[1:]
+
+	if results != nil {
+		process.VoteCount = results.Height
+		if process.Results, err = aggregateResults(meta, results); err != nil {
+			log.Errorf("could not aggregate results: %v", err)
```
Comment (p4u) [inline urlapi/util.go:147]: IMO this method should return an error and all these log.Errorf and some warnings should return (else it will be hard to find bugs here, the logs are hard to follow sometimes).

File: urlapi/util.go [hunk 28]
```diff
@@ -0,0 +1,380 @@
...
+
+	if process.StartDate, err = u.estimateBlockTime(vc.StartBlock); err != nil {
+		log.Warnf("could not estimate startDate at %d: %v", vc.StartBlock, err)
+	}
+
+	if process.EndDate, err = u.estimateBlockTime(vc.EndBlock); err != nil {
+		log.Warnf("could not estimate endDate at %d: %v", vc.EndBlock, err)
+	}
+
+	process.Ok = true
```
Comment (p4u) [inline urlapi/util.go:159]: I don't understand this. There is not any return so the execution will always get here...

File: urlapi/util.go [hunk 29]
```diff
@@ -0,0 +1,380 @@
...
+		return orgPermissionsInfo{}, fmt.Errorf("entity %X does not belong to this integrator", entityID)
+	}
+	return orgPermissionsInfo{
+		integratorPrivKey: integratorPrivKey,
+		entityID:          entityID,
+		organization:      organization,
+	}, nil
+}
+func (u *URLAPI) estimateBlockTime(height uint32) (time.Time, error) {
+	currentHeight, err := u.vocClient.GetCurrentBlock()
```
Comment (p4u) [inline urlapi/util.go:52]: I would add here a big TODO for vocClient: do save the current block in memory and update it once every 10 seconds or so

File: urlapi/util.go [hunk 30]
```diff
@@ -0,0 +1,380 @@
...
+	if err != nil {
+		return time.Time{}, err
+	}
+	diffHeight := int64(height) - int64(currentHeight)
+
+	if diffHeight < 0 {
+		blk, err := u.vocClient.GetBlock(height)
```
Comment (p4u) [inline urlapi/util.go:59]: same for block... that's gonna be very consuming, it won't scale

**Review by p4u — CHANGES_REQUESTED** [review]

File: urlapi/util.go (hunk 30 above)
Comment (NateWilliams2) [reply urlapi/util.go:59]: we can't cache this block, as it's the block at our target height, not the current height. So there's no way to know which block to cache ahead of time. 
This line of code fetches the block only to get the exact timestamp of that block height. We could alternatively either
- estimate block times of heights in the past (like we do for those in the future), rather than getting the exact value
- implement a gateway API call that returns just the timestamp for a given block

File: vocclient/vocclient.go [hunk 31]
```diff
@@ -433,3 +450,58 @@ func (c *Client) CreateProcess(
 	}
 	return p.Process.StartBlock, nil
 }
+
+func (c *Client) RelayVote(signedTx []byte) (string, error) {
+	var err error
+	var resp *api.APIresponse
+	var req api.APIrequest
```
Comment (p4u) [inline vocclient/vocclient.go:457]: `req := api.APIrequest{ Method: .... }`

**Review by p4u — CHANGES_REQUESTED** [review]

File: vocclient/vocclient.go [hunk 32]
```diff
@@ -433,3 +450,58 @@ func (c *Client) CreateProcess(
...
+
+	if len(resp.Nullifier) < 1 {
+		return "", fmt.Errorf("RelayVote: did not revieve nullifier")
+	}
+
+	return resp.Nullifier, nil
+}
+
+func (c *Client) getVocHeight() error {
+	defer c.blockHeight.lock.Lock()
```
[the remark this answers is not in the document]
Comment (emmdim) [reply vocclient/vocclient.go:478]: yeah yeah. Stupid copy paste

File: vocclient/vocclient.go [hunk 33]
```diff
@@ -266,18 +283,18 @@ func (c *Client) AddClaim(censusID string, censusSigner *ethereum.SignKeys, cens
...
 	resp, err := c.pool.Request(req, c.signingKey)
 	if err != nil {
-		return types.HexBytes{}, err
+		return dvoteTypes.HexBytes{}, err
 	}
 	return resp.Root, nil
 }
 
-func (c *Client) AddClaimBulk(censusID string, censusSigners []*ethereum.SignKeys, censusPubKeys []string, censusValues []*types.BigInt) (root types.HexBytes, invalidClaims []int, _ error) {
+func (c *Client) AddClaimBulk(censusID string, censusSigners []*ethereum.SignKeys, censusPubKeys []string, censusValues []*dvoteTypes.BigInt) (root dvoteTypes.HexBytes, invalidClaims []int, _ error) {
```
Comment (p4u) [inline vocclient/vocclient.go:297]: line too long
